### PR TITLE
Implemented dashboard page and handled logout related bug

### PR DESCRIPTION
**Pull request by ShubhamTiwari55**
This PR introduces a new dashboard page to the application, which utilizes the previously implemented ticket management functions to provide a comprehensive interface for viewing and managing tickets. The dashboard includes features for viewing tickets based on different filters, updating ticket statuses, and displaying ticket details. Additionally, access to the dashboard and other protected routes is now restricted to authenticated users only.

**Changes Made**
**Dashboard Page Implementation**

Created a new file src/pages/Dashboard.js to serve as the dashboard page.
Integrated ticket management functions from src/functions/ticketFunctions.js to enable interaction with Firestore.
Implemented UI components to display tickets and allow for ticket status updates and filtering.

**Dashboard Page Components**

Ticket List: Displays a list of tickets with relevant details.
Filters: Allows users to filter tickets by user ID, product, and status.
Status Update: Provides functionality to update the status of selected tickets.

**Firebase Integration**

Utilized Firebase Admin SDK for database interactions.
Ensured that Firestore queries and updates are correctly reflected in the UI.

**Authentication Guard**

Implemented an AuthProvider and useAuth hook in src/context/authContext.js to manage user authentication state.
Created a ProtectedRoute component in src/components/ProtectedRoute.js to restrict access to authenticated users only.
Wrapped protected routes in App.js with ProtectedRoute to ensure only logged-in users can access them.

Screenshot 2024-06-10 193948

**Testing**
Verified that the dashboard page loads correctly and interacts with Firestore.
Tested ticket filtering, viewing, and status updating to ensure proper functionality.
Confirmed that only authenticated users can access the dashboard and other protected routes.

### CODE REVIEW

File: src/Components/Sidebar/Sidebar.jsx [hunk 1]
```diff
@@ -1,18 +1,36 @@
...
 import { ICONS, ACTIVE_ICONS, ACTIVE_CLASS } from '../../constants';
+import { signOut } from 'firebase/auth';
+import { auth } from '../../config/firebase.config';
+import { toast } from 'react-toastify';
+import 'react-toastify/dist/ReactToastify.css';
+import { TOAST_MESSAGES } from '../../constants';
 const Sidebar = () => {
+
+  const navigate = useNavigate();
+  const handleLogout = async () => {
```
Comment (Jha2508) [inline src/Components/Sidebar/Sidebar.jsx:16]: docs

File: src/Components/Sidebar/Sidebar.jsx [hunk 2]
```diff
@@ -1,18 +1,36 @@
...
+    try {
+      await signOut(auth);
+      toast.success(TOAST_MESSAGES.SIGN_OUT_SUCCESS);
+      navigate("/login");
```
Comment (Jha2508) [inline src/Components/Sidebar/Sidebar.jsx]: move this route to a constants route file

File: src/Components/SignInWithGoogle.jsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import React from 'react';
+import { useNavigate } from 'react-router-dom';
+import { signInWithPopup, GoogleAuthProvider } from 'firebase/auth';
+import { auth, db } from '../config/firebase.config';
+import { toast } from 'react-toastify';
+import '../Screens/login.css';
+import { IMAGES } from '../constants';
+import { doc, setDoc } from 'firebase/firestore';
+
+function SignInWithGoogle() {
```
Comment (Jha2508) [inline src/Components/SignInWithGoogle.jsx:13]: docs

File: src/Components/SignInWithGoogle.jsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  const navigate = useNavigate();
+
+  async function googleLogin() {
+    const provider = new GoogleAuthProvider();
+    try {
+      const result = await signInWithPopup(auth, provider);
+      await setDoc(doc(db,"Users",result.user.uid),{
```
Comment (Jha2508) [inline src/Components/SignInWithGoogle.jsx]: fetch the db name from constants

File: src/Components/SignInWithGoogle.jsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        email: result.user.email,
+        username: result.user.displayName,
+      });
+      toast.success("Signed in with Google successfully!", {
```
Comment (Jha2508) [inline src/Components/SignInWithGoogle.jsx]: toast message should be from constants

File: src/Components/Stats.jsx [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import './Stats.css';
+
+const Stats = ({ stats }) => {
```
Comment (Jha2508) [inline src/Components/Stats.jsx:6]: docs

File: src/Components/TicketList.jsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import './TicketList.css';
+
+const TicketList = ({ tickets }) => {
```
Comment (Jha2508) [inline src/Components/TicketList.jsx:6]: docs

File: src/Screens/Dashboard.jsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import React, { useEffect, useState } from 'react';
+import TicketList from '../Components/TicketList';
+import Stats from '../Components/Stats';
+import { viewTickets } from '../utils/networkHelper';
+import './Dashboard.css';
+
+const Dashboard = () => {
```
Comment (Jha2508) [inline src/Screens/Dashboard.jsx:8]: docs

File: src/Screens/login.jsx [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+import '../Screens/login.css';
+import { IMAGES, TOAST_MESSAGES } from '../constants';
+import { useNavigate } from 'react-router-dom';
+import { signInWithEmailAndPassword } from 'firebase/auth';
+import { auth } from '../config/firebase.config';
+import { toast } from 'react-toastify';
+import 'react-toastify/dist/ReactToastify.css';
+import SignInWithGoogle from '../Components/SignInWithGoogle';
+
+const LoginPage = () => {
```
Comment (Jha2508) [inline src/Screens/login.jsx:13]: docs

File: src/Screens/login.jsx [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+  const navigate = useNavigate();
+
+  const handleSubmit = async (e) => {
+    e.preventDefault();
+    try {
+      await signInWithEmailAndPassword(auth, email, password);
+      toast.success(TOAST_MESSAGES.LOGIN_SUCCESS, {
+        position: "top-center",
+      });
+      navigate("/home");
```
Comment (Jha2508) [inline src/Screens/login.jsx]: move to route constants file

File: src/Screens/signUp.jsx [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        await setDoc(doc(db,"Users",user.uid),{
+          email: user.email,
+          username: username,
+        });
+      }
+      //console.log("user registered successfully");
+      toast.success(TOAST_MESSAGES.SIGNUP_SUCCESS,{
+        position: "top-center",
+      });
+      navigate("/home");
```
Comment (Jha2508) [inline src/Screens/signUp.jsx]: reuse the constant made in above comment here

File: src/Components/SignInWithGoogle.jsx [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        position: "top-center",
+      });
+      navigate("/home");
+      console.log(result);
+    } catch (error) {
+      toast.error(error.message, {
+        position: "bottom-center",
```
Comment (Jha2508) [inline src/Components/SignInWithGoogle.jsx:36]: even if we get an error, we should not show the message directly to the user, instead use some generic error text as the error message is for logging purpose not showing to the user

File: src/Components/Sidebar/Sidebar.jsx (hunk 2 above)
Comment (ShubhamTiwari55) [reply src/Components/Sidebar/Sidebar.jsx]: Done

File: src/Components/Sidebar/Sidebar.jsx (hunk 1 above)
Comment (ShubhamTiwari55) [reply src/Components/Sidebar/Sidebar.jsx:16]: Done

File: src/Components/SignInWithGoogle.jsx (hunk 3 above)
Comment (ShubhamTiwari55) [reply src/Components/SignInWithGoogle.jsx:13]: Done

File: src/Components/SignInWithGoogle.jsx (hunk 4 above)
Comment (ShubhamTiwari55) [reply src/Components/SignInWithGoogle.jsx]: Done

File: src/Components/SignInWithGoogle.jsx (hunk 5 above)
Comment (ShubhamTiwari55) [reply src/Components/SignInWithGoogle.jsx]: Done

File: src/Components/SignInWithGoogle.jsx (hunk 12 above)
Comment (ShubhamTiwari55) [reply src/Components/SignInWithGoogle.jsx:36]: Done

File: src/Components/Stats.jsx (hunk 6 above)
Comment (ShubhamTiwari55) [reply src/Components/Stats.jsx:6]: Done

File: src/Components/TicketList.jsx (hunk 7 above)
Comment (ShubhamTiwari55) [reply src/Components/TicketList.jsx:6]: Done

File: src/Screens/Dashboard.jsx (hunk 8 above)
Comment (ShubhamTiwari55) [reply src/Screens/Dashboard.jsx:8]: Done

File: src/Screens/login.jsx (hunk 9 above)
Comment (ShubhamTiwari55) [reply src/Screens/login.jsx:13]: Done

File: src/Screens/login.jsx (hunk 10 above)
Comment (ShubhamTiwari55) [reply src/Screens/login.jsx]: Done

File: src/Screens/signUp.jsx (hunk 11 above)
Comment (ShubhamTiwari55) [reply src/Screens/signUp.jsx]: Done

**Review by Jha2508 — APPROVED** [review]

**Pull request merged**